packages/graph-editor/src/components/controls/array.tsx

```diff
@@ -0,0 +1,178 @@
+import { observer } from 'mobx-react-lite';
+import { JSONTree } from 'react-json-tree';
+import React from 'react';
+import { IField } from './interface';
+import { toJS } from 'mobx';
+import { Button, IconButton, Select, Stack, TextInput } from '@tokens-studio/ui';
+import { Minus, Plus } from 'iconoir-react';
+import { ColorPickerPopover } from '../colorPicker';
+import { ANY, AllSchemas, COLOR, Input, NUMBER, STRING } from '@tokens-studio/graph-engine';
+import { delayedUpdateSelector } from '@/redux/selectors';
+import { useSelector } from 'react-redux';
+
+const inputItemTypes = [STRING, NUMBER, COLOR];
+
+export const ArrayField = observer(({ port }: IField) => {
+    const [value, setValue] = React.useState(port.value);
+    const [selectItemsType, setSelectItemsType] = React.useState(port.type.items.$id);
+
+    const [autofocusIndex, setAutofocusIndex] = React.useState<number | null>(null);
+    const useDelayed = useSelector(delayedUpdateSelector);
+
+    React.useEffect(() => {
+        setValue(port.value);
+    }, [port.value]);
+
+    React.useEffect(() => {
+        if (autofocusIndex !== null) {
+            setAutofocusIndex(null);
+        }
+    }, [autofocusIndex]);
+
+    const portName = port.constructor.name;
+    const itemsType = port.type.items.$id;
+
+    const onColorChange =
+        (e: React.ChangeEvent<HTMLInputElement>, index: number) => {
+            let newColor: string;
+            // Weird  problem with the color picker if the user decides to use the text input
+            if (typeof e === 'string') {
+                newColor = e
+            } else {
+                newColor = e.target.value;
+            }
+
+            onChange(newColor, index);
+        };
+
+    const onChange = (newValue: any, index: number) => {
+        const newValueArray = value.map((val: any, i: number) => {
+            if (i === index) {
+                return itemsType === NUMBER ? +newValue : newValue;
+            }
+            return val;
+        });
+        setValue(newValueArray);
+
+        if (useDelayed) {
+            return;
+        }
+
+        (port as Input).setValue(newValueArray);
+    };
+
+    const addItem = () => {
+        const newValueArray = [...value];
+
+        switch (itemsType) {
+            case STRING:
+                newValueArray.push('');
+                break;
+            case NUMBER:
+                newValueArray.push(undefined);
+                break;
+            case COLOR:
+                newValueArray.push('#000000');
+                break;
+            default:
+                newValueArray.push('');
+                break;
+        }
+
+
+        setValue(newValueArray);
+        (port as Input).setValue(newValueArray);
+
+        setAutofocusIndex(newValueArray.length - 1);
+    };
+
+    const removeItem = (index: number) => {
+        const newValueArray = value.filter((val: any, i: number) => i !== index);
```

Comment:
This is bad, we're iterating through every value in the array, rather just join 2 slices together instead of an O(n) iteration to filter out the index